src/main/java/week6/ParkingLot.java

```diff
@@ -0,0 +1,153 @@
+      //if emptyBox is in right pos, we swap the emptyBox with one of the wrong cars
+      emptyBox = correctEmpty(misplacedCars, target, emptyBox, moves);
+      emptyBox = moveEmpty(misplacedCars, target, emptyBox, moves);
+    }
+
+    logger.info(moves.toString());
+    return moves;
+  }
+
+
+  /*
+   * Updates the map misPlacedCars : cars that needs to be moved -> current position.   *
+   * Returns the index of the empty box
+   *
+   * Time Complexity: O(n) where n is the size of array source
+   * Space Complexity: O(1)
+   *
+   * @param source : the initial configuration of the cars
+   * @param target : the target configuration of the cars
+   * @param misplacedCars : the map to be updated cars to be moved -> their current position
+   * @return : the index of the empty box
+   */
+  private static int populateMap(int[] source, int[] target,
+      HashMap<Integer, Integer> misplacedCars) {
+    int emptyBox = -1;
+    for (int i = 0; i < source.length; i++) {
+      if (source[i] == EMPTY) {
+        emptyBox = i;
+      } else if (source[i] != target[i]) {
+        misplacedCars.put(source[i], i);
+      }
+    }
+    return emptyBox;
+  }
+
+  /*
+  * This steps is required if emptyBox is in the right position but the list is not sorted.
+  *     e.g. cars = [1, 0, 2, 3], target = [3, 0, 2, 1] and emptyBox = 1
+  * We swap the emptyBox with one of the cars in the map (wrongly positioned), here we perform
+  *     "move from 1 to 0"
+  *
+  * Amortized Time Complexity: O(1)
+  * Space Complexity: O(1)
+  *
+  * @param misplacedCars : map to be updated with the cars to be moved
+  * @param target : the target configuration of the cars
+  * @param emptyBox : the index of the emptyBox
+  * @param moves : the list of moves which is appended if emptyBox needs to be swapped
+  * @return : the index of the emptyBox
+  */
+  private static int correctEmpty(Map<Integer, Integer> misplacedCars, int[] target, int emptyBox,
+      List<Move> moves) {
+    if (target[emptyBox] == EMPTY) {
+      Entry<Integer, Integer> wrongPosCar = misplacedCars.entrySet().iterator().next();
+      moves.add(new Move(emptyBox, wrongPosCar.getValue()));
+      return misplacedCars.replace(wrongPosCar.getKey(), emptyBox);
+    }
+
+    return emptyBox;
+  }
+
+  /*
+  * Swaps the emptyBox with the car that should be there according to target and remove
+  * car from map (as it is in correct pos)
+  *
+  * Amortized Time Complexity: O(1)
+  * Space Complexity: O(1)
+  *
+  * @param misplacedCars : map to be updated with the cars to be moved
+  * @param target : the target configuration of the cars
+  * @param emptyBox : the index of the emptyBox
+  * @param moves : the list of moves which is appended if emptyBox needs to be swapped
+  * @return : the index of the emptyBox
+  */
+  private static int moveEmpty(Map<Integer, Integer> misplacedCars, int[] target, int emptyBox,
+      List<Move> moves) {
+    if (misplacedCars.containsKey(target[emptyBox])) {
```

Comment:
It might be worth extracting the `target[emptyBox]` into `currentCar` for better readability but is fine either way